Refactor visualization module

aperezhortal opened the pull request:
# Refactor the visualization module

## 1) Refactor visualization.utils functions 

- parse_proj4_string: A more robust implementation of this function is implemented in pyproj. The refactored implementation uses a direct call to the pyproj function.
- proj4_to_cartopy: Refactor the proj4_to_cartopy function
  - Use the parse_proj4_string to construct a dictionary with the proj definitions
  - Simplify the function code by using additional dictionaries to relate the Pyproj and Cartopy parameters. 
  - Add support for additional cartopy keywords defined in the Proj4 definitions.
- Add two plotting utility functions to avoid code duplication and to improve the axis handling:
  - get_basemap_axis: This function safely gets a basemap axis. If the current axis is not a cartopy axis, it creates a basemap. Otherwise, the current axis is returned.
  - get_geogrid: Returns the x,y grid used in the plots along with additional metadata. 
-  Solves issue #197.

## 2) Refactor plotting functions

- Refactor the plotting functions to use the new plotting utils
- Merge `_plot_field_pcolormesh` and `_plot_field` in a single function.
- Avoid plotting the no-data mask separately. Instead, plot it in the `_plot_field` function. 
- Merge `quiver` and `streamplots` functions into a single one called `plot_motion` for avoiding code duplication.
The `quiver` and `streamplot` functions are now wrappers for `plot_motion`.
- **IMPORTANT**
  - Rename the `type` keyword in  `plot_precip_field` to `ptype` to avoid shadowing the built-in function with the same name. Add the corresponding deprecation warning.

## 3) Define a unique zorder for each type of plot

- Basemap features zorder
  - ocean: 0
  - land: 0
  - lakes: 0
  - rivers_lake_centerlines: 0
  - coastline: 15
  - cultural: 15
  - reefs: 15
  - minor_islands: 15
- precipitation: 10
- quiver: 20
- stream function: 30
- tracks and contours: 40

aperezhortal committed:
Small refactoring

- **kwargs was replaced by the explicit keywords declarations
- If unsupported kwargs are provided, they are ignored and a deprecation warning is shown.

aperezhortal committed:
change variable name: gl -> grid_lines

aperezhortal committed:
Refactor visualization module

- Use a single function to plot the quiver and the streamplots to avoid code duplication. This function is called by the quiver and streamplot functions.
- Use different zorders for each type of plot
- Improve the handling of existing geoaxes
- Add two new utils functions to avoid code repetitions

aperezhortal committed:
Update tests with new ptype keyword.

- Fix bug in plot_precip_field handling the old type keyword.

aperezhortal committed:
Change the `step` keyword order for backwards compatibility.

aperezhortal committed:
Merge `_plot_field_pcolormesh` and `_plot_field`

Also, avoid plotting the no-data mask separately.
Instead, plot it in the `_plot_field` function.

aperezhortal committed:
Improve utils docstrings

aperezhortal committed:
Change variables X,Y for x_grid, y_grid

aperezhortal committed:
Use pyproj builtin functions in pyproj_proj4_string

aperezhortal committed:
Minor docstring edit

aperezhortal committed:
Refactor the proj4_to_cartopy function

- Use the parse_proj4_string for construct a dictionary with the proj definitions
- Simplify the function code by using additional dictionaries to relate the Pyproj and Cartopy parameters.
- Add support for additional cartopy keywords defined in the Proj4 definitions.

aperezhortal committed:
Fix reverted PYPROJ_PROJ_KWRDS_TO_CARTOPY dict

The PYPROJ_PROJ_KWRDS_TO_CARTOPY key,values where reverted (values, keys). This produced incorrect cartopy projections that where unrelated with the proj4 definitions.

aperezhortal committed:
Fix minor code issues

aperezhortal committed:
Fix style issues and remove unused imports

dnerini commented on pysteps/visualization/basemaps.py:
```diff
@@ -94,35 +115,25 @@ def plot_geography(
         crs,
         extent,
-        cartopy_scale,
+        scale,
```
The `scale` argument is very much specific to Cartopy and a different basemap plotting library might use a different convention, that's why the argument was previously named `cartopy_scale`.

dnerini commented on pysteps/visualization/basemaps.py:
```diff
@@ -11,13 +11,13 @@
     plot_geography
     plot_map_cartopy
 """
+from cartopy.mpl.geoaxes import GeoAxesSubplot
```
As Cartopy is an optional dependency, this import should be included in a try ... except ImportError ... statement.

dnerini commented on pysteps/visualization/utils.py:
```diff
@@ -10,10 +10,19 @@
+import warnings
+
 import numpy as np
+from cartopy.mpl.geoaxes import GeoAxesSubplot
```
Here too, Cartopy is an optional dependency

dnerini commented on pysteps/visualization/utils.py:
```diff
@@ -28,6 +37,41 @@
+    "rf": "inverse_flattening",
+}
+
+PYPROJ_PROJECTION_TO_CARTOPY = dict(
```
Similarly to above, this dictionary can only be defined if Cartopy is installed.

dnerini reviewed (commented):
Well, this is really excellent work Andres and a very much needed refactoring of the visualization module! It brings a lot of clarity in the code and solves both long standing and newer issues, thanks a lot for the contribution!

I also still need to test it directly myself with some code and data (including the pysteps examples), but for the moment I went through the PR and left few minor comments.

dnerini commented:
> 
    Basemap features zorder
        ocean: 0
        land: 0
        lakes: 0
        rivers_lake_centerlines: 0
        coastline: 2
        cultural: 2
        reefs: 2
        minor_islands: 2
    precipitation: 10
    quiver: 20
    stream function: 30

I like the way you formalized the zorder conventions used when plotting. However, personally I like to see the national boundaries on top of the precipitation field: this allows to keep a geographical reference even when the precipitation field is very large and covers most of the domain. Hence, I'd suggest to set the coastline, cultural, reefs and minor_islands features with a zorder > 10, what do you think?

aperezhortal commented:
> I'd suggest to set the coastline, cultural, reefs and minor_islands features with a zorder > 10, what do you think?

:+1:  I'll make the changes.

aperezhortal committed:
Import cartopy modules inside try:except

aperezhortal committed:
Rename motion plot function

aperezhortal committed:
Set coastline and cultural zorder to 15

Sets the coastline, cultural, reefs and minor_islands features with a zorder = 15. In this way, they are plotted on top of the precipitation field.

aperezhortal committed:
Fix sphinx warnings

aperezhortal committed:
Improve missing dependencies handling

aperezhortal committed:
Fix quiver matplotlib's url

aperezhortal committed:
Handle error if gcc is installed in OSX

aperezhortal committed:
Install gcc9 if only if it is not installed

dnerini commented:
Hi @aperezhortal, thanks for including my previous comments in this PR.

I had a look at the examples rendered from this branch, and unfortunately, it seems that something's still off, for example in thunderstorm_detection_and_tracking.py:

Figure 1, v:latest:
![image](https://user-images.githubusercontent.com/11967971/104229828-0b13e400-544d-11eb-9e6d-5903fe51cd3a.png)

Figure 1, v:refactor_visualization_module:
![image](https://user-images.githubusercontent.com/11967971/104229885-2252d180-544d-11eb-9112-d096ac0b6795.png)

I haven't had the chance to investigate the cause of this yet, but hopefully it relates to something minor.

aperezhortal committed:
Return geo_grids with the same yorigin as the input data

- If geodata is not None but cartopy is not installed, return the x/y grids in the coordinates units defined in geogrid.

aperezhortal committed:
Improve geogrid=None handling

aperezhortal committed:
Remove unused UnsupportedSomercProjection exception

A workaround for the missing Somerc projection was already implemented in the visualization.utils module:

# Note: ccrs.epsg(2056) doesn't work because the projection
# limits are too strict.
# We'll use the Stereographic projection as an alternative.
  somerc=ccrs.Stereographic,

aperezhortal committed:
Remove unused imports

aperezhortal committed:
Remove unused import

aperezhortal committed:
Update zorder for tracks and contours

aperezhortal committed:
Merge branch 'master' of https://github.com/pySTEPS/pysteps into refactor_visualization_module

aperezhortal committed:
Add autoscale to motion plots

For certain situations, the quiver plots were not tight and a margin was shown

aperezhortal committed:
Fix bug in geo_grid

The y_grid was returned in the incorrect order for yorigin=upper

aperezhortal committed:
Fix the bad-data mask

aperezhortal committed:
Use new ptype keyword

aperezhortal committed:
Check that one of geodata and refshape is not None

aperezhortal committed:
Remove empty line after docstring

aperezhortal commented:
Hi @dnerini, thanks for spotting this! The `get_geogrid` function was returning the wrong coordinates units when cartopy was not installed. I also updated the `thunderstorms` visualization module to deal with the geogrid=None case and remove the catch for the `UnsupportedSomercProjection` exception. A workaround was previously implemented and that exception was never risen.

Now the documentation is almost equal to the main branch. You may note small differences in the motion vectors since now they are plotted using the grid coordinates with the same y-origin of the data. Thus, the "steps" starts on the y-origin passed in the geodata (upper) instead of starting at the bottom as before.

dnerini committed:
Fix transparency when plotting probabilities

dnerini commented:
This looks really great, Andres. I noticed only one last inconsistency with the current master branch, Figure 4 in plot_steps_nowcast.py:

v:latest
![image](https://user-images.githubusercontent.com/11967971/104366072-1fbdae00-5519-11eb-98c0-4a6b779c0587.png)

v:refactor_visualization_module
![image](https://user-images.githubusercontent.com/11967971/104366029-10d6fb80-5519-11eb-9403-7623171ea6e9.png)

Notice the reddish background color in the second image? I think the transparency for zero probabilities is not working properly.
Digging a bit into the code, I figured that `cmap.set_under("white", alpha=0)` was missing when plotting probabilities. It also seems that with `cmap.set_under`, it is not necessary to manually flag all values below the minimum threshold to obtain the desired transparency. 

I implemented my suggested changes in 9a0049c. Feel free to modify or revert them as needed.

aperezhortal committed:
Remove vmin, vmax keywords

Since BoundaryNorm was used, there is no need to set the vmin, vmax explicitly.

dnerini committed:
Refactor animation module

dnerini committed:
Run black

dnerini committed:
Remove unused import; improve docstrings

aperezhortal committed:
Use f-strings

aperezhortal committed:
Add plt show to the examples

plt.show was added to make the scripts running correctly from the command line.
